awscli/customizations/s3/results.py

```diff
@@ -180,49 +184,78 @@ def __call__(self, result):
         self._result_handler_map.get(type(result), self._record_noop)(
             result=result)
 
-    def _get_progress_cache_key(self, result):
+    def _get_cache_key(self, result):
         if not isinstance(result, BaseResult):
             raise ValueError(
-                'Any result using _get_progress_cache_key must subclass from '
+                'Any result using _get_cache_key must subclass from '
                 'BaseResult. Provided result is of type: %s' % type(result)
             )
         return ':'.join([result.transfer_type, result.src, result.dest])
 
     def _pop_result_from_caches(self, result):
-        total_progress = self._progress_cache.pop(
-            self._get_progress_cache_key(result))
-        total_file_size = self._total_size_cache.pop(
-            self._get_progress_cache_key(result))
+        cache_key = self._get_cache_key(result)
+        total_progress = self._progress_cache.pop(cache_key, 0)
+        total_file_size = self._total_size_cache.pop(cache_key, None)
         return total_progress, total_file_size
 
     def _record_noop(self, **kwargs):
         # If the result does not have a handler, then do nothing with it.
         pass
 
     def _record_queued_result(self, result, **kwargs):
-        self._progress_remaining_cache[
-            self._get_progress_cache_key(result)] = result.total_transfer_size
+        total_transfer_size = result.total_transfer_size
+        self._total_size_cache[
+            self._get_cache_key(result)] = total_transfer_size
+        # The total transfer size can be None if we do not know the size
+        # immediately so do not add to the total right away.
+        if total_transfer_size:
+            self.expected_bytes_transferred += total_transfer_size
         self.expected_files_transferred += 1
-        self.expected_bytes_transferred += result.total_transfer_size
 
     def _record_progress_result(self, result, **kwargs):
-        self._progress_remaining_cache[
-            self._get_progress_cache_key(result)] -= result.bytes_transferred
-        self.bytes_transferred += result.bytes_transferred
+        bytes_transferred = result.bytes_transferred
+        self._account_for_updated_total_transfer_size_if_needed(result)
+        self._progress_cache[self._get_cache_key(result)] += bytes_transferred
+        self.bytes_transferred += bytes_transferred
+
+    def _account_for_updated_total_transfer_size_if_needed(self, result):
```

Comment:
How about? ``_account_for_newly_discovered_total_transfer_size()`` Or is that too long?